### PR TITLE
8252889: Obsolete -XX:+InsertMemBarAfterArraycopy

### PR DESCRIPTION
**Pull request by TobiHartmann**
The InsertMemBarAfterArraycopy product flag (enabled by default) is broken. I propose to simply remove it.
The corresponding CSR (JDK-8252890) is still waiting for approval.

Thanks,
Tobias

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8252889](https://bugs.openjdk.java.net/browse/JDK-8252889): Obsolete -XX:+InsertMemBarAfterArraycopy

### Reviewers
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**)
 * [Aditya Mandaleeka](https://openjdk.java.net/census#adityam) (@adityamandaleeka - Author)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/69/head:pull/69`
`$ git checkout pull/69`

### CODE REVIEW

File: src/hotspot/share/runtime/arguments.cpp [hunk 1]
```diff
@@ -553,6 +553,7 @@ static SpecialFlag const special_jvm_flags[] = {
   { "UseNewFieldLayout",             JDK_Version::jdk(15), JDK_Version::jdk(16), JDK_Version::jdk(17) },
   { "UseSemaphoreGCThreadsSynchronization", JDK_Version::undefined(), JDK_Version::jdk(16), JDK_Version::jdk(17) },
   { "ForceNUMA",                     JDK_Version::jdk(15), JDK_Version::jdk(16), JDK_Version::jdk(17) },
+  { "InsertMemBarAfterArraycopy",    JDK_Version::undefined(), JDK_Version::jdk(16), JDK_Version::jdk(17) },
```
Comment (shipilev) [inline src/hotspot/share/runtime/arguments.cpp:556]: This seems like a product flag, so needs to follow `ForceNUMA` example? E.g. 16 -> 17 -> 18? And a provisional CSR?

File: src/hotspot/share/opto/macroArrayCopy.cpp [hunk 2]
```diff
@@ -1101,9 +1101,6 @@ void PhaseMacroExpand::expand_arraycopy_node(ArrayCopyNode *ac) {
     merge_mem = MergeMemNode::make(mem);
     transform_later(merge_mem);
 
-    RegionNode* slow_region = new RegionNode(1);
-    transform_later(slow_region);
-
```
Comment (shipilev) [inline src/hotspot/share/opto/macroArrayCopy.cpp:1106]: This cleanup does not seem related?

**Review by shipilev — CHANGES_REQUESTED** [review]
Synopsis might be changed to "Obsolete -XX:+InsertMemBarAfterArraycopy".

File: src/hotspot/share/opto/macroArrayCopy.cpp (hunk 2 above)
Comment (TobiHartmann) [reply src/hotspot/share/opto/macroArrayCopy.cpp:1106]: Yes but it does not seem justified to create a separate PR for removing dead code in the same file.

File: src/hotspot/share/runtime/arguments.cpp (hunk 1 above)
Comment (TobiHartmann) [reply src/hotspot/share/runtime/arguments.cpp:556]: Since the flag is broken, I think immediate obsoletion, rather than deprecation makes more sense. See also discussion in JDK-8252889. A CSR has been created (JDK-8252890).

Comment (shipilev) [reply src/hotspot/share/runtime/arguments.cpp:556]: Okay then!

File: src/hotspot/share/opto/macroArrayCopy.cpp (hunk 2 above)
Comment (shipilev) [reply src/hotspot/share/opto/macroArrayCopy.cpp:1106]: Okay then.

**Review by shipilev — APPROVED** [review]

**Comment by TobiHartmann** [comment]
Thanks for the review, Aleksey!

**Review by adityamandaleeka — APPROVED** [review]
Not a reviewer, but this looks good to me.

**Comment by TobiHartmann** [comment]
Thanks for the review!
